s3transfer/manager.py

```diff
@@ -378,11 +389,114 @@ def _register_handlers(self):
+
+
+class TransferCoordinatorCanceler(object):
+    def __init__(self):
+        """Abstraction to control the cancelation of all transfers
+
+        This abstraction allows the manager to wait for inprogress transfers
+        to complete and cancel all inprogress transfers.
+        """
+        self._lock = threading.Lock()
+        self._tracked_transfer_coordinators = set()
+
+    @property
+    def tracked_transfer_coordinators(self):
+        """The set of transfer coordinators being tracked"""
+        with self._lock:
+            # We return a copy because the set is mutable and if you were to
+            # iterate over the set, it may be changing in length due to
+            # additions and removals of transfer coordinators.
+            return copy.copy(self._tracked_transfer_coordinators)
+
+    def add_transfer_coordinator(self, transfer_coordinator):
+        """Adds a transfer coordinator of a transfer to be canceled if needed
+
+        :type transfer_coordinator: s3transfer.futures.TransferCoordinator
+        :param transfer_coordinator: The transfer coordinator for the
+            particular transfer
+        """
+        with self._lock:
+            self._tracked_transfer_coordinators.add(transfer_coordinator)
+
+    def remove_transfer_coordinator(self, transfer_coordinator):
+        """Remove a transfer coordinator from cancelation consideration
+
+        Typically, this method is invoked by the transfer coordinator itself
+        to remove its self when it completes its transfer.
+
+        :type transfer_coordinator: s3transfer.futures.TransferCoordinator
+        :param transfer_coordinator: The transfer coordinator for the
+            particular transfer
+        """
+        with self._lock:
+            self._tracked_transfer_coordinators.remove(transfer_coordinator)
+
+    def cancel(self):
+        """Cancels all inprogress transfers
+
+        This cancels the inprogress transfers by calling cancel() on all
+        tracked transfer coordinators.
         """
-        self._submission_executor.shutdown()
-        self._request_executor.shutdown()
-        self._io_executor.shutdown()
+        for transfer_coordinator in self.tracked_transfer_coordinators:
+            transfer_coordinator.cancel()
+
+    def wait(self):
+        """Wait until there are no more inprogress transfers"""
+        try:
+            for transfer_coordinator in self.tracked_transfer_coordinators:
+                transfer_coordinator.result()
+        except KeyboardInterrupt:
+            # If Keyboard interrupt is raised while waiting for
+            # the result, then exit out of the wait and raise the
+            # exception
+            raise
+        except Exception:
+            # A general exception could have been thrown because
+            # of result(). We just want to ignore this and continue
+            # because we at least know that the transfer coordinator
+            # has completed.
+            pass
```

Comment:
That would mean that any result would be propogated in the the context handler. So if I have logic like this:
```py
with manager as m:
     future = m.upload('file', 'fakebucket', 'key')

future.result()
```
The error will actually happen exiting the context manager instead of at the ``future.result()``. This is a larger issue if you are trying to batches of transfers and then want to analyze all the results at the end.

The only reason I am using ``result()`` is because it is the only method on the transfer coordinator that is hooked up to the done event. If you are fine with it, I can add a new ``wait_until_done()`` method on the ``TransferCoordinator`` if that makes sense?